modules/swagger-codegen/src/main/resources/Javascript-Flowtyped/api.mustache

```diff
@@ -0,0 +1,266 @@
+            if ({{paramName}} !== undefined && {{paramName}} !== null) {
+                localVarHeaderParameter['{{baseName}}'] = (({{paramName}}:any):string);
+            }
+            {{/isListContainer}}
+
+    {{/headerParams}}
+    {{#formParams}}
+            {{#isListContainer}}
+            if ({{paramName}}) {
+            {{#isCollectionFormatMulti}}
+                {{paramName}}.forEach((element) => {
+                    localVarFormParams.append('{{baseName}}', element);
+                })
+            {{/isCollectionFormatMulti}}
+            {{^isCollectionFormatMulti}}
+                    localVarFormParams.set('{{baseName}}', {{paramName}}.join(COLLECTION_FORMATS["{{collectionFormat}}"]));
+            {{/isCollectionFormatMulti}}
+            }
+            {{/isListContainer}}
+            {{^isListContainer}}
+            if ({{paramName}} !== undefined) {
+                localVarFormParams.set('{{baseName}}', (({{paramName}}:any):string));
+            }
+            {{/isListContainer}}
+
+    {{/formParams}}
+    {{#bodyParam}}
+            {{^consumes}}
+            localVarHeaderParameter['Content-Type'] = 'application/json';
+            {{/consumes}}
+            {{#consumes.0}}
+            localVarHeaderParameter['Content-Type'] = '{{{mediaType}}}';
+            {{/consumes.0}}
+
+    {{/bodyParam}}
+            localVarUrlObj.query = Object.assign({}, localVarUrlObj.query, localVarQueryParameter, options.query);
+            // fix override query string Detail: https://stackoverflow.com/a/7517673/1077943
+            delete localVarUrlObj.search;
+            localVarRequestOptions.headers = Object.assign({}, localVarHeaderParameter, options.headers);
+    {{#hasFormParams}}
+            localVarRequestOptions.body = localVarFormParams;
+    {{/hasFormParams}}
+    {{#bodyParam}}
+            const needsSerialization = (typeof {{paramName}} !== "string") || localVarRequestOptions.headers['Content-Type'] === 'application/json';
+            localVarRequestOptions.body =  needsSerialization ? JSON.stringify({{paramName}} || {}) : ((({{paramName}}:any):string) || "");
+    {{/bodyParam}}
+
+            return {
+                url: url.format(localVarUrlObj),
+                options: localVarRequestOptions,
+            };
+        },
+    {{/operation}}
+    }
+};
+
+export type {{classname}}Type = { {{#operation}}
+    {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}options?: RequestOptions): Promise<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}Response{{/returnType}}>,
+    {{/operation}}
+}
+
+/**
+ * {{classname}} - factory function to inject configuration {{#description}}
+ * {{{description}}}{{/description}}
+ * @export
+ */
+export const {{classname}} = function(configuration?: Configuration, fetch: FetchAPI = portableFetch): {{classname}}Type {
+    const basePath: string = (configuration && configuration.basePath) || BASE_PATH;
```

Comment:
@jaypea Unless I'm missing something this line makes it so the `basePath` cannot be changed during runtime.

Reply:
@wouterhund thanks for clarification. I see your point now.
I like to have my code as side-effect free as possible and like to treat the objects as immutable as possible. So changing something in `Configuration` should result in a new Configuration object and be passed explicitly to the users of this object.
Ask yourself, how would you unit test your code?

I'd rather rewrap the api whenever the configuration needs to be changed.

my code looks more like that:
```
// apiWrapper.js
export function wrapApi() {
  const myConfig = new apiclient.Configuration({
    apiKey: token,
    basePath: (process.env.REACT_APP_API_BASEPATH: any),
  });
  return { myApi : apiclient.MyApi(myConfig) }
}

// someSaga.js
import { wrapApi } from 'apiWrapper.js';

wrapApi().myApi.aMethodInside(parameter);
```

Reply:
The configuration object itself is not immutable, so I wouldn't expect it to be treated as such.

With your example, every single time I want to invoke the api a new configuration and api object gets instantiated as well, which seems like unnecessary overhead. The function could of course cache the myApi object until it changes, but at that point I'd rather keep using my custom templates which don't suffer from this issue. This looks much more practical to me:

```
// api.js
const configuration = new apiclient.Configuration();

let updateCredentials = () => {
  let state = store.getState().api;

  configuration.basePath = state.url;
  configuration.username = state.user;
  configuration.password = state.pass;
}

store.subscribe(updateCredentials)
updateCredentials()

export const api = apiclient.MyApi(configuration);

// someSaga.js
import { api } from 'api.js';
api.aMethodInside(parameter);
```

Additionally, the regular javascript code generated by swagger-codegen also doesn't try to be immutable. And since I'm migrating from that this was a pain point.